### PR TITLE
VFL with Azure Cache for Redis 

### PR DESCRIPTION
**Pull request by majercakdavid**
## Purpose
Azure Cache for Redis used as a communication channel between the nodes by utilizing Redis streams.

## What is the expected review turnaround time?

Urgency:
- [ ] High (needs review today)
- [X] Medium (needs review within a few days - most common)
- [ ] Low (can wait a week)

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: examples/components/CCFRAUD_VERTICAL/traininsilo/aml_comm.py [hunk 1]
```diff
@@ -350,3 +352,275 @@ def close(self):
...
+            rank (int): rank of the current node
+            world_size (int): total number of nodes
+            run_id (str): run id of the current run
+            connection_string (str, optional): connection string to the Redis server. Defaults to None.
+            message_timeout (int, optional): timeout for the Redis messages. Defaults to 60.
+            connect_timeout (int, optional): timeout for the initial connection to other nodes. Defaults to 1800.
+        """
+        super().__init__(rank, world_size, run_id)
+
+        if connection_string is None or len(connection_string) == 0:
```
Comment (jfomhover) [inline examples/components/CCFRAUD_VERTICAL/traininsilo/aml_comm.py:384]: nit: I believe you can catch both by using `if not connection_string:`

File: docs/tutorials/vertical-fl.md [hunk 2]
```diff
@@ -64,6 +64,22 @@ This can all be performed with ease using a data provisioning pipeline. To run i
...
+
+### Redis streams
+In case it is not feasible to use vnets in your case you can fallback on using Redis. This, however, involves provisioning [Azure Cache for Redis](https://azure.microsoft.com/en-us/products/cache/). Please, make sure that you provision at least P1 Premium Instance with 6GB cache for the demos. However, if you would like to use it for your own data and architectures, feel free to scale it according to the needs for messages being sent between the nodes.
+
+Once you provision *Azure Cache for Redis* (this can be provisioned in whichever subscription you like):
+1. Go to *Access keys* and copy *Primary connection string*. 
+2. Go to your *Azure Machine Learning workspace* in Azure Portal and click on the *Key Vault* item.
+3. In the Key Vault go to *Secrets* and *Generate* new one with previously copied connection string and the following name "amlcomm-redis-connection-string".
+4. Go to both **host.py** and **contributor.py**, add `from aml_comm import AMLCommRedis` at the top of the file and change **AMLCommSocket** to **AMLCommRedis** while keeping the same parameters as for **AMLCommSocket**.
```
Comment (jfomhover) [inline docs/tutorials/vertical-fl.md:80]: can we replace this by a parameter in the components?

File: examples/components/CCFRAUD/traininsilo/run.py [hunk 3]
```diff
@@ -608,7 +608,7 @@ def run(args):
         total_num_of_iterations=args.total_num_of_iterations,
         experiment_name=args.metrics_prefix,
         iteration_name=args.iteration_name,
-        device_id=int(os.environ.get("LOCAL_RANK", "0")),
```
Comment (jfomhover) [inline examples/components/CCFRAUD/traininsilo/run.py:611]: shouldn't we stick to use LOCAL_RANK instead to support multi-node?

File: examples/components/CCFRAUD_VERTICAL/traininsilo/aml_comm.py [hunk 4]
```diff
@@ -350,3 +352,275 @@ def close(self):
...
+    def _get_connection_string(self) -> str:
+        try:
+            from azureml.core import Run, Keyvault
+
+            logger.warning("Getting Redis connection string from Azure ML Key Vault")
+            run = Run.get_context()
+            ws = run.experiment.workspace
+            kv: Keyvault = ws.get_default_keyvault()
+            connection_string = kv.get_secret("amlcomm-redis-connection-string")
+            logger.warning("Got Redis connection string from Azure ML Key Vault")
```
Comment (garg-amit) [inline examples/components/CCFRAUD_VERTICAL/traininsilo/aml_comm.py:458]: nit: `logger.info` ?

File: docs/tutorials/vertical-fl.md [hunk 5]
```diff
@@ -64,6 +64,22 @@ This can all be performed with ease using a data provisioning pipeline. To run i
...
+
+### Sockets
+This is easier to start with as the only **requirement here is that the nodes are interconnected using vnet**. This is also the **default option**. To create this type of communication channel just create instance of a **AMLCommSocket** class with world size, rank of the node and root run id
+
+### Redis streams
+In case it is not feasible to use vnets in your case you can fallback on using Redis. This, however, involves provisioning [Azure Cache for Redis](https://azure.microsoft.com/en-us/products/cache/). Please, make sure that you provision at least P1 Premium Instance with 6GB cache for the demos. However, if you would like to use it for your own data and architectures, feel free to scale it according to the needs for messages being sent between the nodes.
+
+Once you provision *Azure Cache for Redis* (this can be provisioned in whichever subscription you like):
+1. Go to *Access keys* and copy *Primary connection string*. 
+2. Go to your *Azure Machine Learning workspace* in Azure Portal and click on the *Key Vault* item.
```
Comment (garg-amit) [inline docs/tutorials/vertical-fl.md:78]: you might also need to add these instructions https://github.com/Azure-Samples/azure-ml-federated-learning/blob/damajercak/vertical-fl-redis/docs/real-world-examples/ccfraud.md#option-2-using-azure-ui

**Review by jfomhover — APPROVED** [review]
Looks good to me! thank you!

**Comment by jfomhover** [comment]
Let's just wait for the tests to complete

**Review by garg-amit — APPROVED** [review]

**Pull request merged**